sudoku_board.cc

```diff
@@ -1,123 +1,181 @@
-    for (int i = cube_top_r; i < cube_top_r + CUBE_SIZE; i++){
-        for (int j = cube_left_c; j < cube_left_c + CUBE_SIZE; j++){
-            values_in_cube.insert(m_board_values[i][j]);
-        }
+std::set<int> SudokuBoard::calc_forbidden_values_in_cube(size_t r,
+                                                         size_t c) const {
+  std::set<int> values_in_cube{};
+  int cube_top_r = r - (r % CUBE_SIZE);
+  int cube_left_c = c - (c % CUBE_SIZE);
+  for (size_t i = cube_top_r; i < cube_top_r + CUBE_SIZE; i++) {
+    for (size_t j = cube_left_c; j < cube_left_c + CUBE_SIZE; j++) {
+      values_in_cube.insert(m_board_values[i][j]);
     }
-    values_in_cube.erase(EMPTY_CELL_SIGN);
-    return values_in_cube;
+  }
+  values_in_cube.erase(EMPTY_CELL_SIGN);
+  return values_in_cube;
 }
 
-
-std::set<int> SudokuBoard::get_values_in_row(int r, int c){
-    std::set<int> values_in_row(m_board_values[r].begin(), m_board_values[r].end());
-    values_in_row.erase(EMPTY_CELL_SIGN);
-    return values_in_row;
+std::set<int> SudokuBoard::calc_forbidden_values_in_row(size_t r,
+                                                        size_t c) const {
+  std::set<int> values_in_row(m_board_values[r].begin(),
+                              m_board_values[r].end());
+  values_in_row.erase(EMPTY_CELL_SIGN);
+  return values_in_row;
 }
 
-
-std::set<int> SudokuBoard::get_values_in_col(int r, int c){
-    std::set<int> values_in_col{};
-    for (int i = 0; i < m_size; i++){
-        values_in_col.insert(m_board_values[i][c]);
-    }
-    values_in_col.erase(EMPTY_CELL_SIGN);
-    return values_in_col;
+std::set<int> SudokuBoard::calc_forbbiden_values_in_col(size_t r,
+                                                        size_t c) const {
+  std::set<int> values_in_col{};
+  for (size_t i = 0; i < m_size; i++) {
+    values_in_col.insert(m_board_values[i][c]);
+  }
+  values_in_col.erase(EMPTY_CELL_SIGN);
+  return values_in_col;
 }
 
-std::set<int> SudokuBoard::get_forbidden_values(int r, int c){
-    std::set<int> forbidden_values = get_values_in_row(r, c);
-    forbidden_values.merge(get_values_in_col(r, c));
-    forbidden_values.merge(get_values_in_cube(r, c));
-    return forbidden_values;
+std::set<int> SudokuBoard::calc_forbidden_values(size_t r, size_t c) const {
+  std::set<int> forbidden_values = calc_forbidden_values_in_row(r, c);
+  forbidden_values.merge(calc_forbbiden_values_in_col(r, c));
+  forbidden_values.merge(calc_forbidden_values_in_cube(r, c));
+  return forbidden_values;
 }
 
-void SudokuBoard::empty_options(){
-    m_possible_options = std::vector(m_size, std::vector(m_size, std::set<int>({})));
-}
\ No newline at end of file
+void SudokuBoard::update_cell_possible_options(size_t r, size_t c) {
+  std::set<int> current_forbidden_values = calc_forbidden_values(r, c);
+  std::set<int> old_possible_options = get_cell_options(r, c);
+  std::set<int> new_possible_options({});
+  std::set_difference(
```

Comment:
Crazy std magic huh

Reply:
You are building the new options every time even if the values didn't change.
Why not just removing the `current_forbidden_values` from `m_possible_options[r][c]`?